src/Spago/Command/Fetch.purs

```diff
@@ -360,28 +360,75 @@ toAllDependencies = foldl (Map.unionWith (\l _ -> l)) Map.empty
 getGitPackageInLocalCache :: forall a. PackageName -> GitPackage -> Spago (Git.GitEnv a) Unit
 getGitPackageInLocalCache name package = do
   let localPackageLocation = Config.getPackageLocation name (GitPackage package)
-  tempDir <- mkTemp' (Just $ printJson Config.gitPackageCodec package)
-  logDebug $ "Cloning repo in " <> tempDir
-  Git.fetchRepo package tempDir >>= case _ of
-    Left err -> die err
-    Right _ -> do
+  existsGitCache <- FS.exists gitCacheLocation
+  { globallyCached, sourceDir } <-
+    if existsGitCache
+      then do
+        logInfo $ "Using global git cache in " <> gitCacheLocation
+        isBranch <- Git.isBranch { ref: package.ref, path: gitCacheLocation }
+        when isBranch do
+          fetchRes <- Git.fetchRepo package gitCacheLocation
+          case fetchRes of
+            Left err -> die err
+            Right _ -> pure unit
+        pure { globallyCached: true, sourceDir: gitCacheLocation }
+      else do
+        -- Not cached in the global cache, so we download it and then see if it can be cached
+        tempDir <- mkTemp' (Just $ printJson Config.gitPackageCodec package)
+        logDebug $ "Cloning repo in " <> tempDir
+        fetchRes <- Git.fetchRepo package tempDir
+        case fetchRes of
+          Left err -> die err
+          Right _ -> do
+            shouldCache <- not <$> Git.isBranch { ref: package.ref, path: tempDir }
+            if shouldCache
+              then do
+                -- Concurrent threads may have created the directory now,
+                -- especially since fetching can take a long time
+                existsGitCache2 <- FS.exists gitCacheLocation
+                when (not existsGitCache2) do
+                  FS.mkdirp globalGitCache
+                  FS.moveSync { src: tempDir, dst: gitCacheLocation }
+                pure { globallyCached: true, sourceDir: gitCacheLocation }
+              else do
+                pure { globallyCached: false, sourceDir: gitCacheLocation }
```

Comment:
Instead of checking again we can just catch errors - if the move operation fails then other fiber likely put the repo there (we don't have true concurrency though, so I'm not sure how likely that is):
```suggestion
            when shouldCache do
              FS.mkdirp globalGitCache
              try (FS.moveSync { src: tempDir, dst: gitCacheLocation }) >>= case _ of
                Left _err -> logDebug $ "Couldn't move cloned repo to " <> gitCacheLocation
                Right _ -> pure unit
            pure { globallyCached: shouldCache, sourceDir: gitCacheLocation }
```